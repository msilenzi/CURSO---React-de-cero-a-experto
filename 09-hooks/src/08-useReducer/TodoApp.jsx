import { useEffect, useReducer } from 'react'
import SectionTitle from '@UI/SectionTitle'
import todoReducer from './todoReducer'
import TodoList from './TodoList'
import TodoAdd from './TodoAdd'
import TodoResume from './TodoResume'

const initialState = []

function init() {
  return JSON.parse(localStorage.getItem('todos')) ?? []
}

function TodoApp() {
  const [todos, dispatch] = useReducer(todoReducer, initialState, init)

  useEffect(() => {
    localStorage.setItem('todos', JSON.stringify(todos))
  }, [todos])

  function handleAdd(todo) {
    dispatch({ type: 'add', payload: todo })
  }

  function handleDelete(id) {
    dispatch({ type: 'delete', payload: id })
  }

  function handleToggle(id) {
    dispatch({ type: 'toggle', payload: id })
  }

  return (
    <div style={{ minWidth: '420px' }}>
      <SectionTitle text="TodoApp" />
      <TodoResume todos={todos} />
      <TodoAdd handleAdd={handleAdd} />
      <TodoList
        todos={todos}
        handleDelete={handleDelete}
        handleToggle={handleToggle}
      />
    </div>
  )
}

export default TodoApp
