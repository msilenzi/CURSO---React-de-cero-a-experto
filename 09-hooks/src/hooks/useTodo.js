import { useEffect, useReducer } from 'react'
import todoReducer from '08-useReducer/todoReducer'

const initialState = []

function init() {
  return JSON.parse(localStorage.getItem('todos')) ?? []
}

function useTodo() {
  const [todos, dispatch] = useReducer(todoReducer, initialState, init)

  useEffect(() => {
    localStorage.setItem('todos', JSON.stringify(todos))
  }, [todos])

  function handleAdd(todo) {
    dispatch({ type: 'add', payload: todo })
  }

  function handleDelete(id) {
    dispatch({ type: 'delete', payload: id })
  }

  function handleToggle(id) {
    dispatch({ type: 'toggle', payload: id })
  }

  return { todos, handleAdd, handleDelete, handleToggle }
}

export { useTodo }
