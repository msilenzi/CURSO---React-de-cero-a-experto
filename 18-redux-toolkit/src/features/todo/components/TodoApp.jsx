import PropTypes from 'prop-types'
import { useGetTodoByIdQuery, useGetTodosQuery } from '../store/api'
import { useState } from 'react'

function TodoApp() {
  // const { data, isLoading } = useGetTodosQuery()

  const [todoId, setTodoId] = useState(1)
  const { data, isLoading } = useGetTodoByIdQuery(todoId)

  return (
    <main
      style={{
        minHeight: '90svh',
        marginTop: 'calc(5svh - 4rem)',
        textAlign: 'left',
      }}
    >
      <h1>Todo - RTK Query</h1>
      <Loading isLoading={isLoading} />
      {/* <TodosList isLoading={isLoading} todos={data} /> */}
      <Todo
        isLoading={isLoading}
        todo={data}
        onPrev={() => setTodoId((id) => id - 1)}
        onNext={() => setTodoId((id) => id + 1)}
      />
    </main>
  )
}

function Loading({ isLoading }) {
  if (!isLoading) return null
  return <h3>Loading...</h3>
}

function TodosList({ isLoading, todos }) {
  if (isLoading) return null

  return (
    <>
      <ul>
        {todos.map((todo) => (
          <li
            key={todo.id}
            style={{
              color: todo.completed ? 'rgba(255, 255, 255, 0.5)' : 'inherit',
              textDecoration: todo.completed ? 'line-through' : 'none',
            }}
          >
            {todo.title}
          </li>
        ))}
      </ul>
    </>
  )
}

function Todo({ isLoading, todo, onPrev, onNext }) {
  if (isLoading) return null

  return (
    <>
      <pre>{JSON.stringify(todo, null, 2)}</pre>

      <div>
        <button onClick={onPrev} disabled={todo.id === 1}>
          &#8810; Prev
        </button>
        <span style={{ margin: '0 2em' }}>{todo.id}</span>
        <button onClick={onNext}>Next &#8811;</button>
      </div>
    </>
  )
}

Loading.propTypes = {
  isLoading: PropTypes.bool.isRequired,
}

TodosList.propTypes = {
  isLoading: PropTypes.bool.isRequired,
  todos: PropTypes.array,
}

Todo.propTypes = {
  isLoading: PropTypes.bool.isRequired,
  todo: PropTypes.object,
  onPrev: PropTypes.func.isRequired,
  onNext: PropTypes.func.isRequired,
}

export default TodoApp
