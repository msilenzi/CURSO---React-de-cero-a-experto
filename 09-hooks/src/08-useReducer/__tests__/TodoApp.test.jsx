import { render } from '@testing-library/react'
import TodoApp from '08-useReducer/TodoApp'
import { useTodo } from '@Hooks/useTodo'

jest.mock('@Hooks/useTodo.js')

describe('Pruebas en <TodoApp />', () => {
  useTodo.mockReturnValue({
    todos: [
      { id: 1, description: 'Todo #1', done: false },
      { id: 2, description: 'Todo #2', done: true },
    ],
    handleAdd: jest.fn(),
    handleDelete: jest.fn(),
    handleToggle: jest.fn(),
  })

  test('debe coincidir con el snapshot', () => {
    const {container} = render(<TodoApp />)
    expect(container).toMatchSnapshot()
  })
})
