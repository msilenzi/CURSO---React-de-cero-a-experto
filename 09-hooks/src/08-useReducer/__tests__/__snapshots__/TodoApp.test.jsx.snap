// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoApp /> debe coincidir con el snapshot 1`] = `
<div>
  <div
    style="min-width: 420px;"
  >
    <h3
      class="mb-4"
    >
      TodoApp
    </h3>
    <p
      class="mb-2 text-muted"
    >
      Total: 
      2
       - Completed: 
      1
       
      <i>
        (
        50.00
        %)
      </i>
    </p>
    <form
      class="input-group mb-4"
    >
      <input
        aria-describedby="btn-add"
        aria-label="New todo"
        class="form-control form-control-lg form-dark"
        name="description"
        placeholder="New todo"
        type="text"
        value=""
      />
      <button
        class="btn btn-dark"
        id="btn-add"
        type="submit"
      >
        Add
      </button>
    </form>
    <ul
      class="list-group"
    >
      <li
        class="todo-item list-group-item list-group-item-action d-flex justify-content-between align-items-center p-2 "
        style="cursor: pointer;"
      >
        <span
          class=""
        >
          Todo #1
        </span>
        <button
          class="todo-item-delete btn btn-sm btn-outline-danger"
        >
          ✕
        </button>
      </li>
      <li
        class="todo-item list-group-item list-group-item-action d-flex justify-content-between align-items-center p-2 bg-secondary-subtle"
        style="cursor: pointer;"
      >
        <span
          class="text-muted text-decoration-line-through"
        >
          Todo #2
        </span>
        <button
          class="todo-item-delete btn btn-sm btn-outline-danger"
        >
          ✕
        </button>
      </li>
    </ul>
  </div>
</div>
`;
